### PR TITLE
Add postgresql data source check at scaling

### PR DESCRIPTION
**Pull request by azexcy**
Fixes #ISSUSE_ID.

Changes proposed in this pull request:
- Add check

### CODE REVIEW

**Comment by azexcy** [comment]
Checks only if user is super role or have replication privilege, no check create table or schema, because create table or schema privilege verified before

File: shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-dialect/shardingsphere-data-pipeline-postgresql/src/main/java/org/apache/shardingsphere/data/pipeline/postgresql/check/datasource/PostgreSQLDataSourceChecker.java [hunk 1]
```diff
@@ -17,18 +17,55 @@
...
+            PreparedStatement preparedStatement = connection.prepareStatement(SHOW_GRANTS_SQL);
+            preparedStatement.setString(1, metaData.getUserName());
+            try (ResultSet resultSet = preparedStatement.executeQuery()) {
+                resultSet.next();
+                String isSuperRole = resultSet.getString(SUPER_ROLE_NAME);
+                String isReplicationRole = resultSet.getString(REPLICATION_ROLE_NAME);
+                log.info("checkPrivilege: isSuperRole: {}, isReplicationRole: {}", isSuperRole, isReplicationRole);
+                if (StringUtils.equalsIgnoreCase(isSuperRole, "f") && StringUtils.equalsIgnoreCase(isReplicationRole, "f")) {
+                    throw new PipelineJobPrepareFailedException("Source data source is lack of REPLICATION privileges.");
+                }
```
Comment (sandynz) [inline shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-dialect/shardingsphere-data-pipeline-postgresql/src/main/java/org/apache/shardingsphere/data/pipeline/postgresql/check/datasource/PostgreSQLDataSourceChecker.java:64]: Could we not limit role to super role?
If user create a role and grant replication privilege, it should work.

File: shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-dialect/shardingsphere-data-pipeline-postgresql/src/main/java/org/apache/shardingsphere/data/pipeline/postgresql/check/datasource/PostgreSQLDataSourceChecker.java [hunk 2]
```diff
@@ -17,18 +17,55 @@
...
     public void checkPrivilege(final Collection<? extends DataSource> dataSources) {
+        for (DataSource each : dataSources) {
+            checkPrivilege(each);
+        }
+    }
+    
+    private void checkPrivilege(final DataSource dataSource) {
+        try (Connection connection = dataSource.getConnection()) {
+            DatabaseMetaData metaData = connection.getMetaData();
+            PreparedStatement preparedStatement = connection.prepareStatement(SHOW_GRANTS_SQL);
```
Comment (sandynz) [inline shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-dialect/shardingsphere-data-pipeline-postgresql/src/main/java/org/apache/shardingsphere/data/pipeline/postgresql/check/datasource/PostgreSQLDataSourceChecker.java:55]: It's better to put `connection.prepareStatement` in try-resource.

File: shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-dialect/shardingsphere-data-pipeline-postgresql/src/main/java/org/apache/shardingsphere/data/pipeline/postgresql/check/datasource/PostgreSQLDataSourceChecker.java [hunk 3]
```diff
@@ -17,18 +17,55 @@
...
+            preparedStatement.setString(1, metaData.getUserName());
+            try (ResultSet resultSet = preparedStatement.executeQuery()) {
+                resultSet.next();
+                String isSuperRole = resultSet.getString(SUPER_ROLE_NAME);
+                String isReplicationRole = resultSet.getString(REPLICATION_ROLE_NAME);
```
Comment (sandynz) [inline shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-dialect/shardingsphere-data-pipeline-postgresql/src/main/java/org/apache/shardingsphere/data/pipeline/postgresql/check/datasource/PostgreSQLDataSourceChecker.java:60]: If ResultSet of SHOW_GRANTS_SQL has no result, `resultSet.next()` might return false, could we just run `resultSet.getString`?

**Review by sandynz — CHANGES_REQUESTED** [review]
Thanks. Some privileges check snippets could be double checked.

File: shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-dialect/shardingsphere-data-pipeline-postgresql/src/main/java/org/apache/shardingsphere/data/pipeline/postgresql/check/datasource/PostgreSQLDataSourceChecker.java (hunk 1 above)
Comment (sandynz) [reply shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-dialect/shardingsphere-data-pipeline-postgresql/src/main/java/org/apache/shardingsphere/data/pipeline/postgresql/check/datasource/PostgreSQLDataSourceChecker.java:64]: By mistake, `SELECT * FROM pg_roles WHERE rolname = ?` 's parameter is current login user.

**Comment by sandynz** [comment]
Refer to https://www.postgresql.org/docs/current/sql-createuser.html

> CREATE USER is now an alias for [CREATE ROLE](https://www.postgresql.org/docs/current/sql-createrole.html). The only difference is that when the command is spelled CREATE USER, LOGIN is assumed by default, whereas NOLOGIN is assumed when the command is spelled CREATE ROLE.

**Comment by azexcy** [comment]
Refer to https://www.postgresql.org/docs/current/sql-createrole.html
> The INHERIT attribute governs inheritance of grantable privileges (that is, access privileges for database objects and role memberships). It does not apply to the special role attributes set by CREATE ROLE and ALTER ROLE. For example, being a member of a role with CREATEDB privilege does not immediately grant the ability to create databases, even if INHERIT is set; it would be necessary to become that role via [SET ROLE](https://www.postgresql.org/docs/current/sql-set-role.html) before creating a database.

File: shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-dialect/shardingsphere-data-pipeline-postgresql/src/main/java/org/apache/shardingsphere/data/pipeline/postgresql/check/datasource/PostgreSQLDataSourceChecker.java (hunk 1 above)
Comment (sandynz) [inline shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-dialect/shardingsphere-data-pipeline-postgresql/src/main/java/org/apache/shardingsphere/data/pipeline/postgresql/check/datasource/PostgreSQLDataSourceChecker.java:64]: Is it possible to supply suggestion of how to add REPLICATION privilege?

File: shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-dialect/shardingsphere-data-pipeline-postgresql/src/main/java/org/apache/shardingsphere/data/pipeline/postgresql/check/datasource/PostgreSQLDataSourceChecker.java [hunk 4]
```diff
@@ -17,18 +17,57 @@
...
  * PostgreSQL Data source checker.
  */
+@Slf4j
 public class PostgreSQLDataSourceChecker extends AbstractDataSourceChecker {
     
+    private static final String SHOW_GRANTS_SQL = "SELECT * FROM pg_roles WHERE rolname = ?";
+    
+    private static final String SUPER_ROLE_NAME = "rolsuper";
+    
+    private static final String REPLICATION_ROLE_NAME = "rolreplication";
```
Comment (sandynz) [inline shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-dialect/shardingsphere-data-pipeline-postgresql/src/main/java/org/apache/shardingsphere/data/pipeline/postgresql/check/datasource/PostgreSQLDataSourceChecker.java:43]: Could we just inline these string? It's short enough and just used one time.

Or could we change it to xxx_COLUMN_NAME? It's not ROLE NAME.

File: shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-dialect/shardingsphere-data-pipeline-postgresql/src/main/java/org/apache/shardingsphere/data/pipeline/postgresql/check/datasource/PostgreSQLDataSourceChecker.java [hunk 5]
```diff
@@ -17,18 +17,57 @@
...
+            try (ResultSet resultSet = preparedStatement.executeQuery()) {
+                if (!resultSet.next()) {
+                    throw new PipelineJobPrepareFailedException(String.format("No role exists, rolname: %s.", metaData.getUserName()));
+                }
+                String isSuperRole = resultSet.getString(SUPER_ROLE_NAME);
+                String isReplicationRole = resultSet.getString(REPLICATION_ROLE_NAME);
+                log.info("checkPrivilege: isSuperRole: {}, isReplicationRole: {}", isSuperRole, isReplicationRole);
+                if (StringUtils.equalsIgnoreCase(isSuperRole, "f") && StringUtils.equalsIgnoreCase(isReplicationRole, "f")) {
+                    throw new PipelineJobPrepareFailedException(String.format("Source data source is lack of REPLICATION privileges, sql example is `ALTER ROLE \"%s\" REPLICATION;`.",
+                            metaData.getUserName()));
```
Comment (sandynz) [inline shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-dialect/shardingsphere-data-pipeline-postgresql/src/main/java/org/apache/shardingsphere/data/pipeline/postgresql/check/datasource/PostgreSQLDataSourceChecker.java:65]: `sql example is` could be `you could try`

**Review by sandynz — CHANGES_REQUESTED** [review]

**Pull request merged**